### PR TITLE
fix direct upload for all content types

### PR DESCRIPTION
**Pull request by machisuji**
WP [#39126](https://community.openproject.org/projects/openproject/work_packages/39126/activity)

Adds `mime-types` to the frontend to determine the mime-type of most common files.
For other, more uncommon files (e.g. ifc) the content type will default to `application/binary` which will be allowed through the attachment whitelist even if not included during the prepare upload step. Those attachments then will be deleted if their content type is still not valid during the FinishDirectUpload job.

Other attachments whose content type could be determined in the frontend already won't be uploaded to begin with if their content type is not included in the attachments whitelist.

### CODE REVIEW

File: frontend/src/app/components/api/op-file-upload/op-direct-file-upload.service.ts [hunk 1]
```diff
@@ -100,13 +101,16 @@ export class OpenProjectDirectFileUploadService extends OpenProjectFileUploadSer
     };
   }
 
-  public getDirectUploadFormFrom(url:string, file:UploadFile|UploadBlob):Promise<PrepareUploadResult> {
+  public async getDirectUploadFormFrom(url:string, file:UploadFile|UploadBlob):Promise<PrepareUploadResult> {
+    const fileName = file.customName || file.name;
+    const contentType = fileName ? lookup(fileName) || file.type : file.type;
```
Comment (oliverguenther) [inline frontend/src/app/components/api/op-file-upload/op-direct-file-upload.service.ts]: Prefer to use the browser's provided content type, as I'd image that would be more reliable. We can fall back to mime-type if that is empty.

```suggestion
    let contentType = fileName || (fileName && lookup(fileName)) || '';
```

File: frontend/tsconfig.base.json [hunk 2]
```diff
@@ -36,7 +36,8 @@
       "@ng-select/ng-select": ["../node_modules/@ng-select/ng-select/"],
       "core-app/*": ["./app/*"],
       "core-components/*": ["./app/components/*"],
-      "core-vendor/*": ["./vendor/*"]
+      "core-vendor/*": ["./vendor/*"],
+      "path": ["../node_modules/path-browserify"]
```
Comment (oliverguenther) [inline frontend/tsconfig.base.json]: This dependency doesn't seem to be used anymore?

File: frontend/package.json [hunk 3]
```diff
@@ -96,13 +97,15 @@
...
     "mark.js": "^8.11.0",
+    "mime-types": "^2.1.33",
     "moment": "^2.24.0",
     "moment-timezone": "^0.5.26",
     "mousetrap": "~1.6.3",
     "ng-dynamic-component": "^6.0.0",
     "ng2-charts": "^2.3.1",
     "ng2-dragula": "^2.1.1",
     "observable-array": "0.0.4",
+    "path-browserify": "^1.0.1",
```
Comment (oliverguenther) [inline frontend/package.json]: Same here, `path-browserify` doesn't seem to be used?

**Review by oliverguenther — CHANGES_REQUESTED** [review]
Appears to be working fine, but I have some tiny remarks.

File: frontend/src/app/components/api/op-file-upload/op-direct-file-upload.service.ts (hunk 1 above)
Comment (machisuji) [reply frontend/src/app/components/api/op-file-upload/op-direct-file-upload.service.ts]: Makes sense. 👍

File: frontend/tsconfig.base.json (hunk 2 above)
Comment (machisuji) [reply frontend/tsconfig.base.json]: It's needed by `mime-types` for it to work in the frontend. I've added a comment explaining it now.

**Pull request merged**